pandas/tests/computation/test_eval.py

```diff
@@ -441,88 +441,79 @@ def check_compound_invert_op(self, lhs, cmp1, rhs):
     def ex(self, op, var_name="lhs"):
         return f"{op}{var_name}"
 
-    def test_frame_invert(self):
+    @pytest.mark.parametrize(
+        "lhs",
+        [
+            DataFrame(randn(5, 2)),
+            DataFrame(randint(5, size=(5, 2))),
+            DataFrame(rand(5, 2) > 0.5),
+            DataFrame({"b": ["a", 1, 2.0], "c": rand(3) > 0.5}),
+        ],
+    )
+    def test_frame_invert(self, lhs):
         expr = self.ex("~")
+        dtype = lhs.T.dtypes[0]
 
-        # ~ ##
-        # frame
-        # float always raises
-        lhs = DataFrame(randn(5, 2))
-        if self.engine == "numexpr":
-            with pytest.raises(NotImplementedError):
-                result = pd.eval(expr, engine=self.engine, parser=self.parser)
-        else:
-            with pytest.raises(TypeError):
-                result = pd.eval(expr, engine=self.engine, parser=self.parser)
+        error_type = None
+        is_engine_numexpr = self.engine == "numexpr"
 
-        # int raises on numexpr
-        lhs = DataFrame(randint(5, size=(5, 2)))
-        if self.engine == "numexpr":
-            with pytest.raises(NotImplementedError):
+        if dtype == "float64":
```

Comment:
i would rather separate this into 2 tests. e.g. one which has a paramterization and evertything works, and another with is a _error version, where things raise. having these if/then/else if very confusing and an anti-pattern.

same for the changes below.